Sources/Aperture/Aperture.swift

```diff
@@ -1,228 +1,662 @@
+		/// The target type for the current recording
+		private var target: Target? = nil
+		
+		/// Allow consumer to set hooks for various events
+		public var onStart: (() -> Void)?
+		public var onFinish: (() -> Void)?
+		public var onError: ((ApertureError) -> Void)?
+		public var onPause: (() -> Void)?
+		public var onResume: (() -> Void)?
+		
+		public func startRecording(
+			target: Target,
+			options: RecordingOptions
+		) async throws {
+			if self.target != nil {
+				throw ApertureError.recorderAlreadyStarted
 			}
-
-			let audioInput = try AVCaptureDeviceInput(device: audioDevice)
-
-			if session.canAddInput(audioInput) {
-				session.addInput(audioInput)
+			
+			self.target = target
+			self.options = options
+			
+			if target != .audioOnly {
+				guard options.targetId != nil else {
+					throw ApertureError.noTargetProvided
+				}
+			}
+			
+			let streamConfig: SCStreamConfiguration = SCStreamConfiguration()
+			let filter: SCContentFilter?
+			
+			let content: SCShareableContent
+			do {
+				content = try await SCShareableContent.excludingDesktopWindows(false, onScreenWindowsOnly: false)
+			} catch let error as SCStreamError {
+				if error.code == .userDeclined {
+					throw ApertureError.noPermissions
+				}
+				
+				throw ApertureError.couldNotStartStream(error)
+			} catch {
+				throw ApertureError.couldNotStartStream(error)
+			}
+			
+			streamConfig.minimumFrameInterval = CMTime(value: 1, timescale: CMTimeScale(target == .audioOnly ? 1 : options.framesPerSecond))
+			streamConfig.showsCursor = options.showCursor
+			if #available(macOS 15.0, *) {
+				streamConfig.showMouseClicks = options.highlightClicks
+			}
+			
+			if options.recordSystemAudio {
+				streamConfig.capturesAudio = true
+			}
+			
+			if #available(macOS 15, *), let microphoneDeviceId = options.microphoneDeviceId  {
+				streamConfig.captureMicrophone = true
+				streamConfig.microphoneCaptureDeviceID = microphoneDeviceId
+			}
+			
+			switch target {
+			case .screen:
+				guard let display = content.displays.first(where: { display in
+					String(display.displayID) == options.targetId
+				}) else {
+					throw ApertureError.targetNotFound(options.targetId!)
+				}
+				
+				let screenFilter = SCContentFilter(display: display, excludingWindows: [])
+				
+				if let cropRect = options.cropRect {
+					streamConfig.sourceRect = cropRect
+				}
+				
+				if #available(macOS 14.0, *) {
+					streamConfig.width = Int(screenFilter.contentRect.width) * options.scaleFactor
+					streamConfig.height = Int(screenFilter.contentRect.height) * options.scaleFactor
+				} else {
+					streamConfig.width = Int(display.frame.width) * options.scaleFactor
+					streamConfig.height = Int(display.frame.height) * options.scaleFactor
+				}
+				
+				filter = screenFilter
+				break
+			case .window:
+				/// We need to cal this before `SCContentFilter` below otherwise an error is thrown: https://forums.developer.apple.com/forums/thread/743615
```

Comment:
```suggestion
				/// We need to call this before `SCContentFilter` below otherwise an error is thrown: https://forums.developer.apple.com/forums/thread/743615
```